### DISCUSSION TITLE
Enjoying this book

### DISCUSSION THREAD

**Discussion by dukester** [modula3/cm3#806] [General]
"Programming in Modula-3 - An Introduction in Programming with Style"
 by Böszörményi & Weich

This book is my first ever **formal** intro to CS.
I bought my first computer in 1981 - an Apple ][e clone from Taiwan, sporting a z80 cpu as well as the 6502. I learned enough z80 assembly to be dangerous. :) From Byte magazine I think.

Next learned Perl4 because I wanted to crank out my own intranet with a database backend. Got a modem and started "gophering" for resources. Fidonet came in handy.

learned some C; learned some Lisp (newLisp, scheme); tried Oberon Native; like Ulm's Oberon-2 compiler - yet never with any formal training.

Now, here I am digging into Modula-3. The above book is great so far. I would have liked to find Harbison's book to read online somewhere. No chancey!

Looking forward to groking CM3!

**Comment by dukester** [comment]
Another great resource - IMHO - for a Modula-3 noob:
[https://cs.fit.edu/~ryan/m3/tutorial/html/m3.html](https://cs.fit.edu/~ryan/m3/tutorial/html/m3.html)

**Reply by MarcusE1W** [reply]
Today I have found this web site:

https://annas-archive.org/md5/d182b629915d46c838b2d70ec1b22aca

**Comment by dukester** [comment]
In the book there is an example:

```
MODULE Hello EXPORTS Main;
IMPORT SIO;
BEGIN
SIO.PutText("Hello, world!\n")
END Hello.
```
There does not appear to be any m3makefile used!
There does not seem to be a Main.m3 file!
What implementation of m3 are they using?
Can this code also be run with cm3, after changing SIO to IO?

**Reply by mikanystrom-intel** [reply]
I think they are just using their own "simplified" library in SIO.

And sure you can use IO instead.

or to get ahead of the game...

Wr.PutText(Stdio.stdout, "Hello, world!\n");

m3make (quake) isn't part of "The Language" so they are probably intending to leave that to "localization".

But since there's effectively only one implementation of Modula-3, maybe they should have left it in.  But that book is also a bit old.

I found Harbison's more useful, but that was just me.  And the Green Book infinitely more so, even though it's much harder to digest, of course.

**Reply by dukester** [reply]
Those books are way too expensive for a niche language that is almost dead. Sad but true.

At any rate, how would I go about compiling this exact code with CM3?

```
MODULE HelloWorld EXPORTS Main; 
 IMPORT IO;
 BEGIN
   IO.Put("Hello World\n")
 END HelloWorld.
```
HelloWorld.m3 only?

**Reply by mikanystrom-intel** [reply]
Sorry, I only now read your bio and am beginning to understand what you are looking for.  I went through some of the same discovery as you, I am sure.  I got my first Commodore for Christmas in 1982.  But I didn't really learn to program properly until I discovered Modula-3 in graduate school.  (I actually sort of knew---had learned Lisp from the masters---, but had thrown it all away on C and Fortran.)

My advice.. you will not find one book out there that covers a good programming language *and* a good introduction to CS.  Unless we are ok with Scheme.

Modula-3 is an excellent language to get started with when you are "getting serious" with programming.  Whenever I taught intro CS courses, I insisted on using it---at one point our department's intro courses were in Scheme, Modula-3 (my course), and Eiffel.  The students dubbed it "the useless languages sequence"!

But I don't really think I can recommend "Programming in Modula-3" as a great book.  I have a copy but almost never look at it.

I would recommend the following library if you want to get M3 and CS at the same time:

-- Harbison's book (if you can't find it on Amazon I can send you a copy when I am back home in three weeks: I have a pile of them)

-- (here's one out of left field) "Software Development in Pascal" by Sartaj Sahni.  Yes it's old, but if you want to learn how to use *procedural programming* properly, it's actually pretty good.  In particular, the section on ADTs (Abstract Data Types).  This is how everyone codes, say, C these days, but it's kind of not taught, just a part of the lore of programming.  And Pascal is close enough to Modula-3 that if you squint a bit, the code is basically working.

-- Aho, Hopcroft, and Ullman for algorithms (later on move to Knuth)

-- if you're really interested in algorithms, I think there's an edition of Sedgewick in M3.

-- to really understand what programming is about ... Structure and Interpretation of Computer Programs by Abelson and Sussman

https://web.mit.edu/6.001/6.037/sicp.pdf

It's in Lisp, but whatever, it's all the same... (you have to squint a fair bit more though)

I can also recommend "What Computing is All About" by Jan L A van de Snepscheut.

-- the Green Book (for later!) (Systems Programming with Modula-3, edited by Greg Nelson)

**Reply by mikanystrom-intel** [reply]
> Those books are way too expensive for a niche language that is almost dead. Sad but true.

I'll challenge your statement on a few levels.

1. Modula-3 is possibly almost dead, and it has been that way for decades (longer than I have been using it).  It seems to stay in that state, and it's fine with most of us that use it.

2. while almost dead, it's certainly not a niche language!  It can do the jobs of Java, C, Pascal, and Python, all in one, better than all four in most cases.

3. Most importantly, what is your time worth?  If you want to master a new programming language, expect to spend a solid month on it (three years if the language is C++).  A month of hard work!  I don't know, if you are absolutely indigent I will gladly send you copies of all the books I mention.  Otherwise, what's $100 or even $200 or heck even $1,000 compared to your month of efforts? 

> 
> At any rate, how would I go about compiling this exact code with CM3?
> 
> ```
> MODULE HelloWorld EXPORTS Main; 
>  IMPORT IO;
>  BEGIN
>    IO.Put("Hello World\n")
>  END HelloWorld.
> ```
> 
> HelloWorld.m3 only?

Here is how to do it:

make a file called m3makefile, containing

import ("libm3")
implementation ("HelloWorld")
program ("hello")

then stick your code yes in HelloWorld.m3

then run 

cm3

from your prompt

Then you will get an executable somewhere, if on my current system, in ../AMD64_LINUX/hello

just run it

../AMD64_LINUX/hello

**Reply by dukester** [reply]
Thanks!! Got it! Either there exists an actual Main.m3 file or <some-filename.m3> is "declared" as "Main" by MODULE <some-filename> EXPORTS Main;

Is that correct?

Just to add some context! I'm 76 yrs old next month. I'm NOT pursuing a IT career whatsoever - just trying to keep the grey cells from dying off too quickly. Chemo for leukemia not helping matters at all. I liked Oberon-2 but M3 with its standalone compiler is more attractive and versatile. If all else fails, I can always fall back on Free Pascal to tinker around with. Thanks a bunch for your input!

**Reply by mikanystrom-intel** [reply]
> Thanks!! Got it! Either there exists an actual Main.m3 file or <some-filename.m3> is "declared" as "Main" by MODULE EXPORTS Main;
> 
> Is that correct?
> 
> Just to add some context! I'm 76 yrs old next month. I'm NOT pursuing a IT career whatsoever - just trying to keep the grey cells from dying off too quickly. Chemo for leukemia not helping matters at all. I liked Oberon-2 but M3 with its standalone compiler is more attractive and versatile. If all else fails, I can always fall back on Free Pascal to tinker around with. Thanks a bunch for your input!

You are right, you want exactly one module that exports Main.

I see!  Well, you know, an "amateur" is someone who loves what he's doing.  Do enjoy yourself learning.  You might want to look at the Scheme book, then.  It's fun!  (And definitely exercises the gray cells.)  Free, as well!

**Reply by mikanystrom-intel** [reply]
and yeah if your MODULE doesn't have EXPORTS in the title, it by default exports its own interface.  So you could just have "MODULE Main;" and be done with it.  The reason to have a different name for Main is so you don't have lots of files all called Main.m3 but doing different things.  (If that bothers you---it doesn't really bother me.)

**Reply by dukester** [reply]
Got it! Main.m3 doesn't bother me as long as they're segregated in different "project" directories. I'm also messing around with Chicken Scheme. I have the SICP book. I once thought that Eiffel would be "my language". However, the open source implementations are immature with poor docs. Could not find a simple, comprehensive tutorial for it. The quest continues!

BTW, you mentioned (up thread) "quake" in context of m3make. What's this "quake"? Not the game Quake?

**Reply by mikanystrom-intel** [reply]
> Got it! Main.m3 doesn't bother me as long as they're segregated in different "project" directories. I'm also messing around with Chicken Scheme. I have the SICP book. I once thought that Eiffel would be "my language". However, the open source implementations are immature with poor docs. Could not find a simple, comprehensive tutorial for it. The quest continues!
> 
> BTW, you mentioned (up thread) "quake" in context of m3make. What's this "quake"? Not the game Quake?

haha, no.. quake is just the name of the language that m3makefiles are written in.  It predates the game Quake by a couple of years :-)

docs for quake (may be somewhat out of date---check m3makefiles in the cm3 repo for working examples)

https://www.opencm3.net/doc/help/cm3/quake.html

**Reply by dukester** [reply]
Thanks for the link! Never played Quake and couldn't grok what a game engine had to do with M3. LOL Slowly getting the hang of it all. Installing Liberty-Eiffel at the moment. Might fry the grey cells anyway! :D

**Comment by RodneyBates** [comment]
On 2/24/23 22:19, Duke Normandin wrote:
>
> In the book there is an example:
>
> |MODULE Hello EXPORTS Main; IMPORT SIO; BEGIN SIO.PutText("Hello, world!\n") END Hello. |
>
> There does not appear to be any m3makefile used!
>

If there is no file named m3makefile in your src directory,
CM3 will try to infer what to do, just from what files it
finds in there.  I haven't tried to provoke this to see
just how far it will work, but I'm very sure it will work
fine if all you have are same-named interface/module pairs
and one module named Main, or probably, one module named
something else but that exports Main.

Stuff like instantiations might be doubtful, and maybe modules
that export >1 interface.  Definitely not importing interfaces
that are neither in src nor in the standard precompiled libraries
libm3 or libm3core.

For anything very big, even if simple, I usually provide an
m3makefile.  The one place I have seen unexpected trouble without
one is if there are leftover .i3 or .m3 files in src that I didn't
want to be part of what is compiled.

**Reply by dukester** [reply]
Thanks for the clarification. I just cranked out a zsh script that creates a M3 project for me. It asks a few questions then creates required directories and the Main.m3 and m3makefile files AND launches emacs with those files. Slick as snot! LOL

**Comment by RodneyBates** [comment]
On 2/25/23 07:01, Duke Normandin wrote:
>
> Those books are way too expensive for a niche language that is almost dead. Sad but true.
>

The text of the green book was recently made available on line,
after getting the copyright owner(s) consent.

I don't have the URL in front of me, but probably someone here
can help immediately, or I can look for it.

If you really want paper, I have a few copies too.

**Reply by dukester** [reply]
Thanks Rodney! Online would be swell. I'll google for it.

**Reply by dukester** [reply]
Can't find it!

**Comment by olaf-wagner** [comment]
On 25 Feb 2023, at 23:02, Duke Normandin wrote:

> Can't find it!

https://archive.org/details/systems-programming-with-modula-3/page/x/mode/2up

>
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/modula3/cm3/discussions/806#discussioncomment-5110865
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***>




-- 
Olaf Wagner -- elego Software Solutions GmbH -- http://www.elegosoft.com
               Gustav-Meyer-Allee 25 / Gebäude 12, 13355 Berlin, Germany
Geschäftsführer: Olaf Wagner | Sitz: Berlin
Handelregister: Amtsgericht Charlottenburg HRB 77719 | USt-IdNr: DE163214194

**Reply by dukester** [reply]
Is that the "green book" Rodney was talking about? I thought Harbison had written it. I DLed it anyway - Thank You!!
